lib/modules/manager/pip-compile/extract.ts

```diff
@@ -216,19 +220,27 @@ export async function extractAllPackageFiles(
   // This needs to go in reverse order to handle transitive dependencies
   for (const packageFile of [...result].reverse()) {
     for (const reqFile of packageFile.managerData?.requirementsFiles ?? []) {
-      let sourceFile: PackageFile | undefined = undefined;
+      let sourceFiles: PackageFile[] | undefined = undefined;
       if (fileMatches.includes(reqFile)) {
-        sourceFile = lockFileSources.get(reqFile);
+        sourceFiles = lockFileSources.get(reqFile);
       } else if (packageFiles.has(reqFile)) {
-        sourceFile = packageFiles.get(reqFile);
+        sourceFiles = [packageFiles.get(reqFile)!];
       }
-      if (!sourceFile) {
+      if (!sourceFiles) {
         logger.warn(
           `pip-compile: ${packageFile.packageFile} references ${reqFile} which does not appear to be a requirements file managed by pip-compile`,
         );
         continue;
       }
-      sourceFile.lockFiles!.push(...packageFile.lockFiles!);
+      for (const sourceFile of sourceFiles) {
+        let merged = [...packageFile.lockFiles!];
+        for (const lockFile of [...sourceFile.lockFiles!].reverse()) {
+          if (!merged.includes(lockFile)) {
+            merged = [lockFile, ...merged];
+          }
+        }
+        sourceFile.lockFiles = merged;
+      }
```

Comment:
Maybe modern ES6 `Set()` can be used here so this can be condensed to a single line. It will also be the fastest.
https://stackoverflow.com/a/14438954

Reply:
The order is important here, and the `Set` composition functions wouldn't give the correct order.

For instance, in the second example above 'mid1.in' would have `lockFiles: ['mid1.txt', 'final.txt']` and 'mid2.in' would have `lockFiles: ['mid2.txt', 'final.txt']`.  'common.in' needs to end up with `lockFiles: ['common.txt', 'mid1.txt', 'mid2.txt', 'final.txt']` so that they get re-generated in the correct order, but doing a simple `Set.union` would result in `['common.txt', 'mid1.txt', 'final.txt', 'mid2.txt']`.

Reply:
I made an exercise out of this. And it didn't pan out. Though you can see how I would aproach the whole problem.
I would extend lockfiles before pushing them through sorting function. I'm leaving this as curiosity, but seems like a dead end.
 
https://github.com/mbudnek/renovate/compare/pip-compile/dash-r-multiple-files...not7cd:renovate:pip-compile/dash-r-multiple-files